src/common/ceph_crypto.cc

```diff
@@ -73,40 +82,68 @@ ssl_get_thread_id(void)
   memcpy(&ret, &t, sizeof(pthread_t));
   return ret;
 }
-#endif /* not OPENSSL_API_1_1 */
+#endif /* not OPENSSL_VERSION_NUMBER < 0x10100000L */
 
 static void init() {
-#ifndef OPENSSL_API_1_1
-  if (++crypto_refs != 1) {
-    return;
+#if OPENSSL_VERSION_NUMBER < 0x10100000L
+  if (++crypto_refs == 1) {
+    // according to
+    // https://wiki.openssl.org/index.php/Library_Initialization#libcrypto_Initialization
+    OpenSSL_add_all_algorithms();
+    ERR_load_crypto_strings();
+
+    // initialize locking callbacks, needed for thread safety.
+    // http://www.openssl.org/support/faq.html#PROG1
+    CRYPTO_set_locking_callback(&ssl_locking_callback);
+    CRYPTO_set_id_callback(&ssl_get_thread_id);
+
+    OPENSSL_config(nullptr);
   }
 
-  /* According to https://wiki.openssl.org/index.php/Library_Initialization#libcrypto_Initialization */
-  OpenSSL_add_all_algorithms();
-  ERR_load_crypto_strings();
-
-  /* Initialize locking callbacks, needed for thread safety.
-   * http://www.openssl.org/support/faq.html#PROG1
-   */
-  CRYPTO_set_locking_callback(&ssl_locking_callback);
-  CRYPTO_set_id_callback(&ssl_get_thread_id);
-
-  OPENSSL_config(nullptr);
-#endif /* OPENSSL_API_1_1 */
+  // we need to record IDs of all threads calling the initialization in
+  // order to *manually* free per-thread memory OpenSSL *automagically*
+  // allocated in ERR_get_state().
+  // XXX: this solution/nasty hack is IMPERFECT. A leak will appear when
+  // a client init()ializes the crypto subsystem with one thread and then
```

Comment:
> init()ializes

typo?

Reply:
Nope, the parentheses have been put to emphasize the method name.